### PR TITLE
[anxpro]  [bittrex] [bitbay] [bleutrade] [cex] [bitmarket] [bitstamp] [btcmarkets] [coinfloor] [coinmate] [exmo] [hitbtc2] [itbit] [therock] updates

### PR DESCRIPTION
**Pull request by npomfret**
anxpro: 
  * v3 api endpoints
  * v2 fetchMyTrades
  * v3 fetchTransations
  * v3 fetchOrders

bittrex: 
  * v3 api endpoints
  * fixed timestamp in deposits
  * replaced fetchClosedOrders with v3 api

bitbay:
 * added new api endpoints
 * implemented fetchTransactions
 * implemented fetchMarkets

bleutrade
 * added doc link
 * v3 added fetchMyTrades 
__needs attention__

bitmarket
 * implemented fetchMarkets
 * added fetchMyTrades
 * added fetchWithdrawals

cex
 * fixed sign method so parameters on most private API calls aren't ignored
 * added fetchOrders (fetchClosedOrder was buggy so started again - needs merging together)
 * added trade parsing to orders

bitstamp
 * handle nonce error

btcmarkets
 * fixed sign method to handle v2 get requests
 * added fetchTransactions

coin floor
 * added fetchMyTrades
 * added fetchTransactions

coinmate
 * added missing api calls
 * added fetchMyTrades
 * added fetchTransactions

exam
 * added support for multiple symbols being passed to fetchMyTrades

hitbtc2:
 * added transaction fee where available

itbit
 * fixed cost calc in orders
 * added average field to orders
 * implemented fetchTransactions

therock
 * fixed sign method so it works with GET requests

### CODE REVIEW

File: js/bittrex.js [hunk 1]
```diff
@@ -49,10 +49,12 @@ module.exports = class bittrex extends Exchange {
                     'account': 'https://{hostname}/api',
                     'market': 'https://{hostname}/api',
                     'v2': 'https://{hostname}/api/v2.0/pub',
+                    'v3': 'https://api.bittrex.com/v3',
                 },
                 'www': 'https://bittrex.com',
                 'doc': [
                     'https://bittrex.github.io/api/',
+                    'https://bittrex.github.io/api/v3',
```
Comment (kroitor) [inline js/bittrex.js]: I thought, maybe we better just detach it into a separate file?

Comment (kroitor) [reply js/bittrex.js]: On a second thought, this may be a way to migrate seamlessly as well, though...

Comment (npomfret) [reply js/bittrex.js]: I think we're going to need a mix of v2 and v3 in this case anyway

**Comment by npomfret** [comment]
@kroitor is there something I need to be carful of when doing block comments?

> ./python/ccxt/async_support/anxpro.py:223:9: E265 block comment should start with '# '

**Comment by kroitor** [comment]
> @kroitor is there something I need to be carful of when doing block comments?

Yes, there's one known bug with commented closing brackets, which i didn't have the time to resolve yet:

https://github.com/ccxt/ccxt/pull/5089/files#diff-31daa752ca9ffe23da14609406478faaR221

```JavaScript
    async someMethod (...) {
        // {
        //     'foo': 'bar',
        // } // ←----- it fails here, due to a chain of substitutions for the py-version
        return foobar;
    }
```

In order to workaround it, just add more tabulation or whitespaces after the comment `//` and before the commented closing bracket

```JavaScript
    async someMethod (...) {
        //     {
        //         'foo': 'bar',
        //     } // ←----- should work
        return foobar;
    }
```

I know this is completely not obvious, but I hope to fix this bug in transpile.js this week.

**Comment by npomfret** [comment]
So I need to indent the comments by 4 spaces to fix?

**Comment by kroitor** [comment]
> So I need to indent the comments by 4 spaces to fix?

Yep, just need more spacing after the opening `//`. Sorry for the inconvenience, I'll fix that bug soon, I hope.

File: js/anxpro.js [hunk 2]
```diff
@@ -80,6 +120,293 @@ module.exports = class anxpro extends Exchange {
...
+        //             processed: '1551357057000',
+        //             timestampMillis: '1557441435956',
+        //             displayTitle: 'Order Fill',
+        //             displayDescription: 'Buy BTC @ 3008.53930 EUR/BTC' } ],
+        //         count: ...,
+        //     timestamp: '1557441435971',
+        //     resultCode: 'OK' }
+        if (response.resultCode !== 'OK')
+            throw new ExchangeError (this.id + ' trade list failed ' + this.json (response));
+        const data = response.transactions.filter (t => t.transactionType === 'WITHDRAWAL' || t.transactionType === 'DEPOSIT');
```
Comment (kroitor) [inline js/anxpro.js]: We can't do closures, so we have a `this.filterBy` base method for these purposes.

Comment (kroitor) [reply js/anxpro.js]: Also, there's a key/property conflict, it has to be `response['transactions']` (single-quoted keys, same everywhere throughout the file: https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#accessing-dictionary-keys)

```JavaScript
const transactions = this.safeValue (response, 'transactions', []);
const groupedTransactions = this.groupBy (transactions, 'transactionType');
const deposits = this.safeValue (groupedTransactions, 'DEPOSIT');
const withdrawals = this.safeValue (groupedTransactions, 'WITHDRAWAL');
const depositsAndWithdrawals = this.arrayConcat (deposits, withdrawals);
```

File: js/bittrex.js (hunk 1 above)
Comment (npomfret) [reply js/bittrex.js]: ... and v1.1!

**Comment by npomfret** [comment]
@kroitor a little help with this build error please...

`Parse error: syntax error, unexpected 'in' (T_STRING), expecting ',' or ')' in /Users/nickpomfret/projects/ccxt.fork/php/bittrex.php on line 829
`

line 829 of the generated code is:

`        if (is_array ($order || 'Type' in $order) && array_key_exists ('OrderType', $order || 'Type' in $order))
`

which comes from :

`        if ('OrderType' in order || 'Type' in order)
`

**Comment by kroitor** [comment]
Yeah, need more brackets )

```JavaScript
if (('OrderType' in order) || ('Type' in order))
```

**Comment by npomfret** [comment]
I'm done with these two exchanges for now, feel free to merge.  I wanted to do more with bittrex but the new API was disappointing.  I've raised some feature requests on their GitHub, fingers crossed they will make some improvements.

**Comment by kroitor** [comment]
I'll add a few minor edits for formatting and stuff, and will merge soon. Thx for the help once again!

**Comment by npomfret** [comment]
Looks like the bittrex API should be improving soon https://github.com/Bittrex/bittrex.github.io/issues/64

**Comment by npomfret** [comment]
@kroitor another build error I don't understand...

This js line:

    url += '/' + this.implodeParams (path, params) + '?' + this.urlencode (query);

Is being compiled to this broken python:

    url += '/' + self.implode_params(path, params) + ' + self.urlencode(query)

It's especially weird because the exact same line is being correctly cross compiled just a few lines before.

**Comment by kroitor** [comment]
@npomfret i think it is falling under the ternary regex, due to the `'?'`... I'll fix this as well as soon as i get to it, hopefully tomorrow.

For now I suggest splitting the line:

```
url += '/' + this.implodeParams (path, params);
if (Object.keys (query)) {
    url += '?' + this.urlencode (query);
}
```

**Comment by npomfret** [comment]
Tried that, same problem.

**Comment by kroitor** [comment]
> Why does it work and then not work?

I can't tell before I get to it, we'll see and fix this.

File: js/gemini.js [hunk 3]
```diff
@@ -230,9 +230,6 @@ module.exports = class gemini extends Exchange {
         }
         let price = this.safeFloat (order, 'price');
         let average = this.safeFloat (order, 'avg_execution_price');
-        if (average !== 0.0) {
-            price = average; // prefer filling (execution) price over the submitted price
-        }
```
Comment (kroitor) [inline js/gemini.js]: If you can split this into separate PRs, I could merge some of these changes rather quickly ) Thx! )

File: js/hitbtc2.js [hunk 4]
```diff
@@ -943,6 +943,27 @@ module.exports = class hitbtc2 extends hitbtc {
         //         address: '0xd53ed559a6d963af7cb3f3fcd0e7ca499054db8b',
         //     }
         //
+        //     {
+        //       "id": "4d241e54-6546-496c-b97c-52178938179d",
```
Comment (kroitor) [inline js/hitbtc2.js]: This comment seems to be duplicating one of the above, no?

Comment (npomfret) [reply js/hitbtc2.js]: Just wanted to show examples of `bankToExchange` and `exchangeToBank` which are internal transfers on the exchange.

Comment (kroitor) [reply js/hitbtc2.js]: Yep, I mean, those are included there (a few lines above) already, no? )

File: js/gemini.js (hunk 3 above)
Comment (npomfret) [reply js/gemini.js]: It's difficult for me to managed multiple PR's right now but I could do it when I'm done

Comment (kroitor) [reply js/gemini.js]: @npomfret 

> when I'm done

that would be awesome! )

thx again!

Comment (frosty00) [reply js/gemini.js]: git checkout branch file )
